### PR TITLE
feat(#463): created admin booking req table

### LINKED ISSUE #463: As an Admin, I can see and filter through a list of booking requests
### User story
As an admin, I need to be able to see a list of all bookings and filter through them.
### Narrative
As an admin I should be able to see a list of bookings and have the ability to filter through it. Unlike approvers that only get to see pending bookings, I should have the ability to filter by status (pending/approved/denied, date range (past/future), and by assignee. 
### Out of scope
This issue doesn't cover the actual process of assigning/reassigning a booking review.
### Acceptance criteria

| # | Given | When | Then |
|---|-------|------|------|
|1|I am logged in as an admin|I navigate to the bookings list|I can see a list of future pending bookings by default and filter options|
|2|I am logged in as an admin|I change the filter options|the bookings list is updated to reflect my chosen filter options (status/date/assigned user)|

### PR DESCRIPTION
**Pull request by jonathanp16**
##### _What existing problem does the pull request solve?_

Created an admin table to view and have the ability to filter by assignee, status and date created. Also have a text input filter to further filter if necessary. If admins want or need to find booking requests with these parameters, this problem is solved.

image
image
image

##### _List all issues this PR resolves_
- closes #463 

##### _Complete these items before requesting reviews_
- [x] Static Analysis Rules Validated
- [x] All commits follow naming conventions

### Sub-tasks Checklist
- [x] UI Implementation
- [x] UI Test
- [x] Backend Implementation
- [x] Backend Tests
- [ ] Integration Tests
- [ ] System Tests
- [x] Documentation

### CODE REVIEW

File: routes/api.php [hunk 1]
```diff
@@ -23,9 +24,11 @@
 
     Route::name('bookings.reviews.assignable')->middleware(['permission:bookings.approve'])
         ->get('/bookings/assignable', [BookingReviewController::class, 'assignable']);
-
 });
 
+
+Route::post('/filterBookingRequests', [BookingRequestController::class, 'filter']);
```
Comment (alexstojda) [inline routes/api.php:30]: This should be behind the `auth:sanctum' guard. @a-drew has a fix incoming for this

Comment (jonathanp16) [reply routes/api.php:30]: Yes I figured there was some sort of authentication needed, pushed this so we could discuss this in the PR. @a-drew will this fix be up by tonight/is it already up?

**Review by alexstojda — CHANGES_REQUESTED** [review]
Nice work, but requesting that auth guard fix be applied before merging

File: routes/api.php (hunk 1 above)
Comment (a-drew) [reply routes/api.php:30]: the fix is already in main since my #507 PR

File: resources/js/Components/Tables/AdminBookingRequestTable.vue [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        date_range_end: null,
+        data_reviewers: []
+      }
+    }
+  },
+  mounted(){
+    const status_zip = {};
+    for (var i = 0; i < this.statuses.length; i++) {
+      status_zip[this.statuses[i]] = false;
+    }
```
Comment (a-drew) [inline resources/js/Components/Tables/AdminBookingRequestTable.vue:194]: haven't seen this type of setup before, I like it. I was using reduce before..

File: tests/Unit/FilterBookingRequestsTest.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+class FilterBookingRequestsTest extends TestCase
+{
+    use RefreshDatabase;
+    /**
+     * Test filtering booking request with status
+     *
+     * @return void
+     */
+    public function test_filter_by_status(){
+        $this->withoutExceptionHandling();
```
Comment (a-drew) [inline tests/Unit/FilterBookingRequestsTest.php:19]: This line should be removed in the final test

File: tests/Unit/FilterBookingRequestsTest.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        $response
+            ->assertStatus(200);
+
+        $res_id = collect($response->json())->pluck('id');
+        $this->assertContains($booking->id, $res_id);
+
+        $response2 = $this->actingAs($user)->postJson('/api/filterBookingRequests', ['status_list' => ['review' => false]]);
+        $response2
+            ->assertStatus(200);
+        $res_id_2 = collect($response2->json())->pluck('id');
```
Comment (a-drew) [inline tests/Unit/FilterBookingRequestsTest.php:40]: you should also perform a `$response->assertSessionHasNoErrors()` after each 200 check.
it helps us catch flaky tests later on

File: app/Http/Controllers/BookingRequestController.php [hunk 5]
```diff
@@ -268,4 +270,54 @@ function ($attribute, $value, $fail) use ($request) {
...
+     */
+    public function filter(Request $request)
+    {
+        // None of the request fields are mandatory, only
+        // filter the ones provided from request
+        $request->validate([
+            'status_list.*' => ['boolean'],
+            'date_range_start' => ['string'],
+            'date_range_end' => ['string'],
+            'data_reviewers' => ['array']
```
Comment (a-drew) [inline app/Http/Controllers/BookingRequestController.php:287]: double check if you'd want any of these to be a required field. I don't think so but worth a comment

**Review by a-drew — CHANGES_REQUESTED** [review]
could of small improvements to make and that more improtant one that alex mentioned with the authentication guard in `api.php`

File: app/Http/Controllers/BookingRequestController.php (hunk 5 above)
Comment (jonathanp16) [reply app/Http/Controllers/BookingRequestController.php:287]: Don't want any of these as a required field. If the user accidentally click on filter without having changed a filter first, there should not be an error.

File: tests/Unit/FilterBookingRequestsTest.php (hunk 4 above)
Comment (jonathanp16) [reply tests/Unit/FilterBookingRequestsTest.php:40]: On it.

File: resources/js/Components/Tables/AdminBookingRequestTable.vue (hunk 2 above)
Comment (jonathanp16) [reply resources/js/Components/Tables/AdminBookingRequestTable.vue:194]: Yes similar to the python zip function implementation, tried to replicate in JS

File: app/Http/Controllers/BookingRequestController.php (hunk 5 above)
Comment (a-drew) [reply app/Http/Controllers/BookingRequestController.php:287]: yeah I figured that was the idea. sounds good, closing convo

**Review by a-drew — APPROVED** [review]
Thanks for covering the changes, great job on the filters once again 🎉 

got some issues on `BookingRequest::59`

**Review by alexstojda — APPROVED** [review]
LGTM

**Review by alecadub — APPROVED** [review]
I have nothing more to say than @alexstojda and @a-drew , 🚢 🇮🇹

**Review by a-drew — APPROVED** [review]
just finished rebasing. tests are passing again and conflicts are resolved. good to merge now 👍

**Review by KeeganWoods1 — APPROVED** [review]
lgtm!

**Pull request merged**